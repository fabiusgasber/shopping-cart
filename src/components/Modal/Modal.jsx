import { useState } from "react";
import styles from "./Modal.module.css"
import PropTypes from "prop-types";

const Modal = ({ title = "Product List", initialShow = false, items=[], handleChange, deleteItem }) => {

    const [open, setOpen] = useState(initialShow);

    const total = items.reduce((prev, curr) => prev + curr.price * curr.quantity, 0).toFixed(2);

    return (
        <>
{
           open && 
           <>
           <div aria-hidden="true" className={styles.backdrop} onClick={() => setOpen(false)}></div>
           <dialog open className={styles.dialog}>
            <button autoFocus aria-label="Close modal" onClick={() => setOpen(false)}>X</button>
            <table className={styles.table}>
                <caption>{title}</caption>
                <tbody>
             {items && items.map(item => 
                (
                    <tr key={item.id} tabIndex={1}>
                        <td><img src={item.image} alt="product"/></td>
                        <td id={styles.title}>{item.title}</td>
                        <td className={styles.price}>{item.price} $</td>
                        <td>
                            <select onChange={(e) => handleChange(e, item, parseInt(e.target.value))} defaultValue={item.quantity}>
                            { Array.from({length: 10}, (_, i) => {
                               return <option key={i}>{i + 1}</option>
                            }) 
                            }
                            </select>
                        </td>
                        <td><button aria-label="remove button" className={styles.removeBtn} type="button" onClick={() => deleteItem(item)}>X</button></td>
                    </tr>
                )
            ) }
                </tbody>
                <tfoot>
                <tr>
                    <th>Total:</th>
                    <td className={styles.price}>{total} $</td>
                </tr>
                </tfoot>
            </table>
            </dialog> 
            </>
}           
        <button className={styles.modalBtn} onClick={() => setOpen(true)} aria-live="polite" aria-atomic>
           <img className="icons" src="/shopping-bag.png" alt="shopping cart"/>
           {items.length > 0 && <p id={styles.cartItemsNumber}>{items.reduce((prev, curr) => prev + curr.quantity, 0)}</p> }
        </button>
        </>
          
    )
}

Modal.propTypes = {
    title: PropTypes.string,
    elem: PropTypes.element,
    initialShow: PropTypes.bool,
    items: PropTypes.array,
    handleChange: PropTypes.func,
    deleteItem: PropTypes.func
}

export default Modal;