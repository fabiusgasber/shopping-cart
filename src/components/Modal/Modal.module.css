body:has(dialog[open]) {
    overflow: hidden;
}

.backdrop {
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    background: rgba(0, 0, 0, 0.5);
}

.dialog {
    display: flex;
    flex-direction: column;
    position: fixed;
    top: 0;
    right: 0;
    left: auto;
    bottom: auto;
    margin: 0;
    padding: 0.5rem 1rem;
    width: min(500px, 50%);
    height: 100vh;
    overflow: auto;
    border: none;
    z-index: 1;
    }

    .dialog[open]{
        animation: fadeIn 0.2s ease-in forwards;
    }

    @keyframes fadeIn {
        from{
            transform: translateX(300px);
        }
        to {
            transform: translateX(0);
        }
    }

.dialog>button {
    align-self: flex-end;
    font-size: 1.2rem;
}

.table {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 2rem;
    text-align: center;
}

.table tbody {
    flex: 1;
}

.table tbody tr {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(80px, 1fr));
    justify-items: center;
    align-items: center;
    margin-bottom: 2rem;
}

#title {
    flex-basis: 100px;
}

.table caption {
    font-size: 1.2rem;
    font-weight: bold;
}

.table img {
    max-width: 50px;
    height: auto;
}

.table tfoot {
    align-self: center;
    font-size: 1.2rem;
}

.table tfoot tr {
    display: flex;
    gap: 1rem;
}

.removeBtn {
    padding: 0.25rem 0.5rem;
    font-size: 1.2rem;
}

.removeBtn:hover {
    background-color: red;
    color: white;
}

.modalBtn {
  display: flex;
  justify-content: center;
  cursor: pointer;
  transition: transform 0.2s ease-in-out;
  position: relative;
}

.modalBtn:hover {
    transform: translateY(-3px);
}

.modalBtn:active {
    transform: translateY(3px);
}

.price {
    font-weight: bold;
    color: var(--secondary-color);
}

#cartItemsNumber {
    position: absolute;
    top: 0;
    right: -10px;
    color: var(--secondary-color);
    font-weight: bold;
    background-color: red;
    color: white;
    border-radius: 50%;
    width: 20px;
    height: 20px;
    text-align: center;
}