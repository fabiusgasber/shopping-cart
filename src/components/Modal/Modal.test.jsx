import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Modal from "./Modal";

describe("Modal component", () => {

    it("renders modal", () => {
        const { container } = render(<Modal />);
        expect(container).toMatchSnapshot();
    })

    it("opens when clicked", async () => {
        const user = userEvent.setup();
        render(<Modal />);
        const modalBtn = screen.getByRole("button", {name: "shopping cart"});

        await user.click(modalBtn);

        expect(screen.getByRole("dialog")).toBeInTheDocument();
    })

    it("is closed when not clicked", () => {
        render(<Modal />)

        expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    })

    it("displays correct title", () => {
        render(<Modal initialShow={true} title={"Shopping Cart"}/>);
        expect(screen.getByText("Shopping Cart")).toBeInTheDocument();
    })

});
