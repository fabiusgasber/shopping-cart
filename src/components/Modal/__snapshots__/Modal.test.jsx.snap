// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component > renders modal 1`] = `
<div>
  <button
    class="_modalBtn_f57fcc"
  >
    <img
      alt="shopping cart"
      class="icons"
      src="/shopping-bag.png"
    />
  </button>
</div>
`;
